File: array/15.py
class Solution(object):
    def threeSum(self, nums):
        res = []
        if len(nums)<3 :
            return res
        nums=sorted(nums)
        for i in range (len(nums)-2):
            if nums[i]>0:
                return res 
            if i>0 and nums[i]==nums[i-1]:
                continue
            l,r=i+1,len(nums)-1
            while(l<r):
                if nums[i]+nums[l]+nums[r]==0:
                    res.append([nums[i],nums[l],nums[r]])
                    while nums[l]==nums[l+1] and l<r :
                        l+=1
                    while nums[r]==nums[r-1] and l<r :
                        r-=1
                    l+=1
                    r-=1
                elif nums[i]+nums[l]+nums[r]>0:
                    r-=1
                else :
                    l+=1
        return res
                  
def main():
    c=Solution()
    print(c.threeSum([-1,0,1,2,-1,-4]))

if __name__ == '__main__':
    main()

File: array/15.js
var threeSum = function(nums) {
  let res = [];
  if (nums.length<3){
    return res;
  }
  nums.sort();
  for (let i=0;i<nums.length-2;i++){
    if(nums[i]>0){
      return res;
    }
    if(i>0 && nums[i]==nums[i-1]){
      continue
    }
    let l=i+1,r=nums.length-1;
    while(l<r) {
      if (nums[i]+nums[l]+nums[r]==0){
        res.push([nums[i],nums[l],nums[r]])
        while(nums[l]==nums[l+1] && l<r) {
          l++
        }
        while(nums[r]==nums[r-1] && l<r) {
          r--
        }
        l++
        r--
      }else if(nums[i]+nums[l]+nums[r]>0){
        r--
      }else{
        l++
      }

    }
  }
  return res;
};

console.log(threeSum([-1,0,1,2,-1,-4]))